gym_wrapper/base_gym_env.py

```diff
@@ -0,0 +1,195 @@
+
+        # specify observation space(s)
+        spaces = {
+            "rgb": gym.spaces.Box(
+                0,
+                255,
+                (
+                    self.game.get_screen_height(),
+                    self.game.get_screen_width(),
+                    self.game.get_screen_channels(),
+                ),
+                # TODO not always true, but in most cases, yes...
+                dtype=np.uint8,
+            )
+        }
+
+        if self.depth:
+            spaces["depth"] = gym.spaces.Box(
+                0,
+                255,
+                (
+                    self.game.get_screen_height(),
+                    self.game.get_screen_width(),
+                ),
+                dtype=np.uint8,
+            )
+        if self.labels:
+            spaces["labels"] = gym.spaces.Box(
+                0,
+                255,
+                (
+                    self.game.get_screen_height(),
+                    self.game.get_screen_width(),
+                ),
+                dtype=np.uint8,
+            )
+
+        self.num_game_variables = self.game.get_available_game_variables_size()
+        if self.num_game_variables > 0:
+            spaces["gamevariables"] = gym.spaces.Box(
+                np.finfo(np.float32).min,
+                np.finfo(np.float32).max,
+                (self.num_game_variables,),
+                dtype=np.float32
+            )
+
+        self.observation_space = gym.spaces.Dict(spaces)
+
+    def step(self, action):
+        assert self.action_space.contains(action), f"{action!r} ({type(action)}) invalid"
+        assert self.state is not None, "Call `reset` before using `step` method."
+
+        # convert action to vizdoom action space (one hot)
+        act = [0 for _ in range(self.action_space.n)]
+        act[action] = 1
+
+        reward = self.game.make_action(act, self.frame_skip)
+        self.state = self.game.get_state()
+        done = self.game.is_episode_finished()
+
+        return self.__collect_observations(), reward, done, {}
+
+    def reset(
+        self,
+        *,
+        seed: Optional[int] = None,
+        return_info: bool = False,
+        options: Optional[dict] = None,
+    ):
+        if seed is not None:
+            self.game.set_seed(seed)
+        self.game.new_episode()
+        self.state = self.game.get_state()
+
+        if not return_info:
+            return self.__collect_observations()
+        else:
+            return self.__collect_observations(), {}
+
+    def __collect_observations(self):
+        observation = {}
+        if self.state is not None:
+            # Check for GRAY8 (no channel dim)
+            screen_buffer = self.state.screen_buffer
+            if screen_buffer.ndim == 2:
+                # Add extra dimension for channel
+                screen_buffer = screen_buffer[None]
+            observation["rgb"] = np.transpose(screen_buffer, (1, 2, 0))
```

Comment:
If we force RGB format then transpose operation will not be needed, additionally, this will produce the wrong output in case of a different than default CRCGCB (CHANNELS, HEIGHT, WIDTH) shape.

Reply:
My brain has been blown: I have always used CRCGCB for RGB images and had fun doing transposes, only to realize now that RGB24 would have solved that :exploding_head: